lib/kernel/src/inet_dns_tsig.erl

```diff
@@ -0,0 +1,340 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright Ericsson AB 2023. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+-module(inet_dns_tsig).
+
+%% DNS TSIG handling
+%%
+%% RFC 8945: Secret Key Transaction Authentication for DNS (TSIG)
+
+%% WARNING: in the spirit of inet_dns.erl, this module only handles the
+%% cryptographic operations and does not absolve you of your responsibities
+%% that include:
+%%  * implement a truncation policy as per RFC8945, section 5.2
+%%  * verifying that for TSIG over TCP the last message contains a TSIG RR
+%%  * truncated messages must include the TSIG RR as per RFC8945, section 5.3
+%%  * if running as a server, generate suitable error responses
+%%    as per RFC8945, section 5.2
+
+-export([init/1, sign/2, sign/3, verify/3]).
+
+-include("inet_dns.hrl").
+
+-define(ALGS_SUPPORTED, [md5, sha, sha224, sha256, sha384, sha512, sha512]).
+-define(ALG_DEFAULT, sha256). % RFC8945, section 6
+
+-define(FUDGE_DEFAULT, 300). % RFC8945, section 10
+
+-define(MAC_SIZE_MIN, 10). % RFC8945, section 5.2.2.1
+-define(TTL, 0).
+
+-define(NOW, os:system_time(seconds)). %  RFC8945, section 5.2.3
+
+-type tsig_key() :: {string(),iodata()}.
+
+-type tsig_algs() :: crypto:hmac_hash_algorithm() | gss_tsig.
```

Comment:
The `crypto:` type is not exported so this causes a Dialyzer error.
Although maybe the `crypto:` type should be supported, I do not think it should be used here, instead enumerate the supported algorithms.  I think 
``` erlang
-type tsig_algs() :: ?ALGS_SUPPORTED.
```
might be what we want.

Btw, `gss_tsig`; how on earth could that even be supported?  Seems complicated.

Reply:
> The `crypto:` type is not exported so this causes a Dialyzer error. Although maybe the `crypto:` type should be supported, I do not think it should be used here, instead enumerate the supported algorithms. I think
> 
> ```erlang
> -type tsig_algs() :: ?ALGS_SUPPORTED.
> ```
> 
> might be what we want.

Sure.

Resolved in https://github.com/erlang/otp/pull/6985/commits/af2995b2f3c6db178c34a076856c83497569546d
 
> Btw, `gss_tsig`; how on earth could that even be supported? Seems complicated.

I did look into it, if you want to bleed from the eyes, head on over to https://stackoverflow.com/a/55936492

From what I can tell the dnspython people avoided the GSSAPI component by having the end user create it OOB and reference it.

Overall, I suspect no one wants OTP to depend on KRB5 to support DNS packets :)

Cheers

Reply:
...spent too long on this staring at the obvious, Dialyzer does not understand Erlang so fixed the fix.

https://github.com/jimdigriz/otp/commit/c1b83588632c499ba6cc251b68cbbbe7e155371e